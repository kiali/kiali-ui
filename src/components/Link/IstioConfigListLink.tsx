import * as React from 'react';
import { Link } from 'react-router-dom';
import { Paths } from '../../config';
import { FilterSelected } from '../Filters/StatefulFilters';

interface Props {
  namespaces: string[];
  errors?: boolean;
  warnings?: boolean;
}

class IstioConfigListLink extends React.Component<Props> {
  namespacesToParams = () => {
    let param: string = '';
    if (this.props.namespaces.length > 0) {
      param = 'namespaces=' + this.props.namespaces.join(',');
    }
    return param;
  };

  validationToParams = () => {
    let params: string = '';

    if (this.props.warnings) {
      params += 'configvalidation=Warning';
    }

    let errorParams: string = '';
    if (this.props.errors) {
      errorParams += 'configvalidation=Not+Valid';
    }

    if (params !== '' && errorParams !== '') {
      params += '&';
    }

    params += errorParams;

    return params;
  };

  cleanFilters = () => {
    FilterSelected.resetFilters();
  };

  render() {
    let params: string = this.namespacesToParams();
    const validationParams: string = this.validationToParams();
    if (params !== '' && validationParams !== '') {
      params += '&';
    }
    params += validationParams;

    return (
      <Link to={`/${Paths.ISTIO}?${params}`} onClick={this.cleanFilters}>
        {this.props.children}
      </Link>
    );
  }
}

export default IstioConfigListLink;
