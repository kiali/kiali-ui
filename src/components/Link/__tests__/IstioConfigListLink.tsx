import * as React from 'react';
import IstioConfigListLink from '../IstioConfigListLink';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

const mockLink = (namespaces: string[], errors?: boolean, warnings?: boolean) => {
  return shallow(<IstioConfigListLink namespaces={namespaces} warnings={warnings} errors={errors} />);
};

describe('Link with only namespaces', () => {
  describe('but only one namespace', () => {
    it('renders the link with only one namespace', () => {
      const wrapper = mockLink(['bookinfo']);

      expect(shallowToJson(wrapper)).toBeDefined();
      expect(shallowToJson(wrapper)).toMatchSnapshot();

      expect(wrapper.find('Link').props().to).toEqual('/istio?namespaces=bookinfo');
    });
  });

  describe('and more than one', () => {
    it('renders the link with only one namespace', () => {
      const wrapper = mockLink(['bookinfo', 'istio-system', 'runtimes']);

      expect(shallowToJson(wrapper)).toBeDefined();
      expect(shallowToJson(wrapper)).toMatchSnapshot();

      expect(wrapper.find('Link').props().to).toEqual('/istio?namespaces=bookinfo,istio-system,runtimes');
    });
  });
});

describe('Link with validation filters', () => {
  describe('and only warnings', () => {
    it('renders link with namespaces and only warning filters', () => {
      const wrapper = mockLink(['bookinfo'], false, true);

      expect(shallowToJson(wrapper)).toBeDefined();
      expect(shallowToJson(wrapper)).toMatchSnapshot();

      expect(wrapper.find('Link').props().to).toEqual('/istio?namespaces=bookinfo&configvalidation=Warning');
    });
  });

  describe('and only errors', () => {
    it('renders link with namespaces and only errors filters', () => {
      const wrapper = mockLink(['bookinfo', 'runtimes'], true, false);

      expect(shallowToJson(wrapper)).toBeDefined();
      expect(shallowToJson(wrapper)).toMatchSnapshot();

      expect(wrapper.find('Link').props().to).toEqual('/istio?namespaces=bookinfo,runtimes&configvalidation=Not+Valid');
    });
  });

  describe('and both errors and warnings', () => {
    it('renders link with namespaces and only errors filters', () => {
      const wrapper = mockLink(['bookinfo', 'runtimes'], true, true);

      expect(shallowToJson(wrapper)).toBeDefined();
      expect(shallowToJson(wrapper)).toMatchSnapshot();

      expect(wrapper.find('Link').props().to).toEqual(
        '/istio?namespaces=bookinfo,runtimes&configvalidation=Warning&configvalidation=Not+Valid'
      );
    });
  });
});
