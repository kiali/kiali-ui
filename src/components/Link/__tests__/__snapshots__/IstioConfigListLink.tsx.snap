// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link with only namespaces and more than one renders the link with only one namespace 1`] = `
<Link
  onClick={[Function]}
  to="/istio?namespaces=bookinfo,istio-system,runtimes"
/>
`;

exports[`Link with only namespaces but only one namespace renders the link with only one namespace 1`] = `
<Link
  onClick={[Function]}
  to="/istio?namespaces=bookinfo"
/>
`;

exports[`Link with validation filters and both errors and warnings renders link with namespaces and only errors filters 1`] = `
<Link
  onClick={[Function]}
  to="/istio?namespaces=bookinfo,runtimes&configvalidation=Warning&configvalidation=Not+Valid"
/>
`;

exports[`Link with validation filters and only errors renders link with namespaces and only errors filters 1`] = `
<Link
  onClick={[Function]}
  to="/istio?namespaces=bookinfo,runtimes&configvalidation=Not+Valid"
/>
`;

exports[`Link with validation filters and only warnings renders link with namespaces and only warning filters 1`] = `
<Link
  onClick={[Function]}
  to="/istio?namespaces=bookinfo&configvalidation=Warning"
/>
`;
